src/crimson/net/SocketConnection.h

```diff
@@ -30,7 +31,7 @@ class SocketConnection;
 using SocketConnectionRef = seastar::shared_ptr<SocketConnection>;
 
 class SocketConnection : public Connection {
-  SocketMessenger& messenger;
+  seastar::weak_ptr<SocketMessenger> messenger;
```

Comment:
It's just not correct to allow a connection to outlive its messenger, other components (e.g. Protocol) are also broken with this change.

The issue you described in the commit message is a side-effect caused by the issue that monc is not aware that the OSD(with its messenger) is already shutdown.

Anyway, the interface `Messenger::connect()` will be changed to throw error if it is already shutdown, according to https://trello.com/c/sYPLRAqB.

Reply:
@cyx1231st Um...actually, the problem that I saw is this: when monc was trying to reconnect to monitors, the OSD hadn't been shutdown. Instead, it is the monitor's shutdown that leads to monc's reconnect. But the reconnect would get into a infinite trying loop, and when the OSD is shutdown, an broken_promise is thrown in the reconnect code path which is leaked from monc to the old daggling connection. Based on your comment [here](https://github.com/ceph/ceph/pull/34492#discussion_r406284744), I think I can fix this problem in the following way:

1. first, make SocketMessenger's shutdown wait for those daggling/wandering connections to close;
2. second, make sure monc leaks no expected exception into the net module

How do you like it?:-)

Reply:
> Instead, it is the monitor's shutdown that leads to monc's reconnect. But the reconnect would get into a infinite trying loop ...

If you are talking about the specific logic https://github.com/ceph/ceph/blob/master/src/crimson/mon/MonClient.cc#L557, I think the monitor should respect the fact that it is stopped and don't reopen new session. This would fix the deadloop properly and also prevent further connect attempts from a potentially shutdown messenger.

> first, make SocketMessenger's shutdown wait for those daggling/wandering connections to close;

I still think it is an error/bug if the application(osd/monc) shuts the messenger down and then tries to establish further connections from it. Messenger should simply deny this attempt, and as a side-effect it also gets rid of the dangling connection issue because:
1. all the existing connection tracked by this messenger will be closed cleanly when the shutdown future is fulfilled, and it is already guaranteed that messenger won't be destructured before that.
2. no more new connection can be created after shutdown.
    * messenger stops accepting new connections when shutdown.
    * messenger refuses to establish new connections if shutdown (**not yet implemented**, partially implemented in your commit "crimson/net: fix messenger shutdown error", and also proposed by https://trello.com/c/sYPLRAqB).

Reply:
> > Instead, it is the monitor's shutdown that leads to monc's reconnect. But the reconnect would get into a infinite trying loop ...
> 
> If you are talking about the specific logic https://github.com/ceph/ceph/blob/master/src/crimson/mon/MonClient.cc#L557, I think the monitor should respect the fact that it is stopped and don't reopen new session. This would fix the deadloop properly and also prevent further connect attempts from a potentially shutdown messenger.

I think OSDs have no way knowing whether the hole cluster is shutting down or only the monitor is rebooting, so OSDs have to try to reconnect so that when monitor come up again, they can establish connections.

> 
> > first, make SocketMessenger's shutdown wait for those daggling/wandering connections to close;
> 
> I still think it is an error/bug if the application(osd/monc) shuts the messenger down and then tries to establish further connections from it. Messenger should simply deny this attempt, and as a side-effect it also gets rid of the dangling connection issue because:
> 
> 1. all the existing connection tracked by this messenger will be closed cleanly when the shutdown future is fulfilled, and it is already guaranteed that messenger won't be destructured before that.
> 2. no more new connection can be created after shutdown.
>    
>    * messenger stops accepting new connections when shutdown.
>    * messenger refuses to establish new connections if shutdown (**not yet implemented**, partially implemented in your commit "crimson/net: fix messenger shutdown error", and also proposed by https://trello.com/c/sYPLRAqB).

Yes, we definitely have to prevent any further connection establishment after the OSD starts the shutdown procedure, but, in cases mentioned above, there could be wandering connections that are created before the shutdown starts, what I mean is we have to wait for these wandering connections to close before shutting down the messenger.

Reply:
> I think OSDs have no way knowing whether the hole cluster is shutting down or only the monitor is rebooting, so OSDs have to try to reconnect so that when monitor come up again, they can establish connections.

In our code a messenger is only shutdown in `OSD::stop()` right? And `monc->stop()` is called before messengers are shutdown. So `monc` actually knows that it is stopped, and it should not try to establish further connections when it is already stopped.

> there could be wandering connections that are unregistered from messengers before the shutdown starts but not yet fully closed when the shutdown happens.

Got your point, yeah, `Messenger::shutdown()` needs to wait for the closing connections that are still pending but already unregistered, it is a missing part in the current implementation. My suggestion is to add a `seastar::gate` at messenger to wait for all the closing operations ended, or, to track all the pending closing connections explicitly and wait for them to be fully closed at the end of shutdown(). The goal is to make sure no more connections are alive after messenger shutdown future is fulfilled.

Reply:
> In our code a messenger is only shutdown in `OSD::stop()` right? And `monc->stop()` is called before messengers are shutdown. So `monc` actually knows that it is stopped, and it should not try to establish further connections when it is already stopped.

Oh, by monitor, I really mean MONs, not monc. The issue came up when I was doing a full cluster shutdown during which monitors are shutdown before OSDs.

> Got your point, yeah, `Messenger::shutdown()` needs to wait for the closing connections that are still pending but already unregistered, it is a missing part in the current implementation. My suggestion is to add a `seastar::gate` at messenger to wait for all the closing operations ended, or, to track all the pending closing connections explicitly and wait for them to be fully closed at the end of shutdown(). The goal is to make sure no more connections are alive after messenger shutdown future is fulfilled.

I'll try to follow your suggestion:-)

Reply:
> Oh, by monitor, I really mean MONs, not monc. The issue came up when I was doing a full cluster shutdown during which monitors are shutdown before OSDs.

But you also said the real issue happens until **when the OSD is shutdown** causing an broken_promise somewhere because the `monc` is still trying to reconnect. I meant `monc` needs to stop trying as soon as it is stopped by OSD. Does it make sense?

Reply:
> But you also said the real issue happens until **when the OSD is shutdown** causing an broken_promise somewhere because the `monc` is still trying to reconnect.

Yes, it should stop trying to reconnect after it is stopped by OSD. But, the problem now is monc tries to reconnect to monitor after monitors are shutdown but before OSDs' shutdown starts, in which case its reconnect is inevitable and would fall into the "connect" loop and would get a broken_promise exception, when messengers shutdown, which is further leaked to the daggling connection the reset event of which triggered the monc's reconnect and further caused the problem. 

>  I meant `monc` needs to stop trying as soon as it is stopped by OSD. Does it make sense?
Yes, absolutely. I think we are saying the same thing from different angles, once we deal with the "broken_promise" exception in the scope of monc, it will stop try to reconnect.

Reply:
> ... would fall into the "connect" loop and would get a broken_promise exception ...

Do you mean "broken_promise exception" happens before you shutdown the OSD, or after? Which promise are you talking about specifically?

Reply:
> > ... would fall into the "connect" loop and would get a broken_promise exception ...
> 
> Do you mean "broken_promise exception" happens before you shutdown the OSD, or after? Which promise are you talking about specifically?

reconnect happens after shutting down the Monitor before shutting down the OSD, "broken_promise exception" happens after shutting down the OSD.

Reply:
> > > ... would fall into the "connect" loop and would get a broken_promise exception ...
> > 
> > 
> > Do you mean "broken_promise exception" happens before you shutdown the OSD, or after? Which promise are you talking about specifically?
> 
> reconnect happens after shutting down the Monitor before shutting down the OSD, "broken_promise exception" happens after shutting down the OSD.

Just checked the monc code quickly, I think the issue is synthesized. The root issue could be that `Client::ms_handle_reset()` is linked with future of `reopen_session()`, which should not. 

Because `reopen_session()` further calls `Connection::authenticate_v2()` and link with `Connection::auth_done.get_future()`. If the monitor is shutdown, this future can never be fulfilled because there will be no replies. In Xuehan's case that messenger fails to track the closing connection, it ends up with "broken_promise". But if messenger can track that connection correctly during shutdown, I think the program will wait for the connection gate for `Client::ms_handle_reset()` forever.

@tchaikov would you mind to take a look?

Reply:
> > > > ... would fall into the "connect" loop and would get a broken_promise exception ...
> > > 
> > > 
> > > Do you mean "broken_promise exception" happens before you shutdown the OSD, or after? Which promise are you talking about specifically?
> > 
> > 
> > reconnect happens after shutting down the Monitor before shutting down the OSD, "broken_promise exception" happens after shutting down the OSD.
> 
> Just checked the monc code quickly, I think the issue is synthesized. The root issue could be that `Client::ms_handle_reset()` is linked with future of `reopen_session()`, which should not.
> 
> Because `reopen_session()` further calls `Connection::authenticate_v2()` and link with `Connection::auth_done.get_future()`. If the monitor is shutdown, this future can never be fulfilled because there will be no replies. In Xuehan's case that messenger fails to track the closing connection, it ends up with "broken_promise". But if messenger can track that connection correctly during shutdown, I think the program will wait for the connection gate for `Client::ms_handle_reset()` forever.

Yes, I think so

> 
> @tchaikov would you mind to take a look?
> 
> @xxhdx1985126 if my understanding is correct, a possible work-around now of this specific issue is to shutdown OSDs before shutdown monitors.

@cyx1231st I think maybe we can set Connection::auth_done with an exception for daggling connections when shutting down messengers, do you think this is feasible?

Reply:
> 2. no more new connection can be created after shutdown.
>    
>    * messenger stops accepting new connections when shutdown.

It seems that, in the current implementation, FixedCPUServerSocket is destroyed in SocketMessenger::shutdown, so I think there wouldn't be no accepting new connections after when shutdown, right?

Reply:
> @cyx1231st @tchaikov I think maybe we can set Connection::auth_done with an exception for connections that hasn't finished the authentication phase when shutting down messengers, do you think this is feasible?

see https://github.com/ceph/ceph/blob/master/src/crimson/mon/MonClient.cc#L434, `monc` already sets the value of promise `auth_done` when stop and when the `active_con` is present. Maybe it is easier to see if there are further problems regarding to the promise, after messenger can track the closing connections correctly.

> @cyx1231st It seems that, in the current implementation, FixedCPUServerSocket is destroyed in SocketMessenger::shutdown, so I think there would be no accepting new connections when shutdown, right?

Yes.

Reply:
> > @cyx1231st @tchaikov I think maybe we can set Connection::auth_done with an exception for connections that hasn't finished the authentication phase when shutting down messengers, do you think this is feasible?
> 
> see https://github.com/ceph/ceph/blob/master/src/crimson/mon/MonClient.cc#L434, `monc` already sets the value of promise `auth_done` when stop and when the `active_con` is present. Maybe it is easier to see if there are further problems regarding to the promise, after messenger can track the closing connections correctly.
> 
> > @cyx1231st It seems that, in the current implementation, FixedCPUServerSocket is destroyed in SocketMessenger::shutdown, so I think there would be no accepting new connections when shutdown, right?
> 
> Yes.

Ok:-)